examples/cifar10/pt/learners/cifar10_autofedrl_learner.py

```diff
@@ -0,0 +1,414 @@
+        if abort_signal.triggered:
+            return make_reply(ReturnCode.TASK_ABORTED)
+        self.log_info(fl_ctx, f"val_acc_local_model: {acc:.4f}")
+
+        # save model
+        self.save_model(is_best=False)
+        if acc > self.best_acc:
+            self.best_acc = acc
+            self.save_model(is_best=True)
+
+        # compute delta model, global model has the primary key set
+        local_weights = self.model.state_dict()
+        model_diff = {}
+        for name in global_weights:
+            if name not in local_weights:
+                continue
+            model_diff[name] = local_weights[name].cpu().numpy() - global_weights[name]
+            if np.any(np.isnan(model_diff[name])):
+                self.system_panic(f"{name} weights became NaN...", fl_ctx)
+                return make_reply(ReturnCode.EXECUTION_EXCEPTION)
+
+        # build the shareable
+        dxo = DXO(data_kind=DataKind.WEIGHT_DIFF, data=model_diff)
+        if hps.get("aw") is not None:
+            # when search aggregation weights, we have to override it
+            # to 1, since we will manually assign weights to aggregator.
+            # Search space will discover which client is more informative.
+            # It might not be related to the number of data in a client.
+            dxo.set_meta_prop(MetaKey.NUM_STEPS_CURRENT_ROUND, 1)
+        else:
+            dxo.set_meta_prop(MetaKey.NUM_STEPS_CURRENT_ROUND, epoch_len)
+
+        self.log_info(fl_ctx, "Local epochs finished. Returning shareable")
+        return dxo.to_shareable()
+
+    def local_valid(self, valid_loader, abort_signal: Signal, tb_id=None, fl_ctx=None, get_loss=False):
+        self.model.eval()
+        with torch.no_grad():
+            correct, total = 0, 0
+            for _, (inputs, labels) in enumerate(valid_loader):
+                if abort_signal.triggered:
+                    return None
+                inputs, labels = inputs.to(self.device), labels.to(self.device)
+                outputs = self.model(inputs)
+                loss = self.criterion(outputs, labels)
+                _, pred_label = torch.max(outputs.data, 1)
+
+                if get_loss:
+                    # return val loss instead of accuracy
+                    # over the number batches
+                    total += inputs.data.size()[0]
+                    correct += loss.item()
+                else:
+                    total += inputs.data.size()[0]
+                    correct += (pred_label == labels.data).sum().item()
+            metric = correct / float(total)
+            if get_loss:
+                self.log_info(fl_ctx, f"HP Search loss: {metric} of {total} batches on {fl_ctx.get_identity_name()}")
+            if tb_id:
+                self.writer.add_scalar(tb_id, metric, self.current_round)
+        return metric
+
+    def validate(self, shareable: Shareable, fl_ctx: FLContext, abort_signal: Signal) -> Shareable:
```

Comment:
I see there is a lot of duplication here VS cifar10_learner.py.

It would be good if we can factor out the common part so we don't need to have same codes at 2 different files.

Reply:
Good suggestion. @guopengf, could you have a look if that can be done?